File: 10. MaxArm Serial Communication/3. Control Example/3.2 Arudino Control Routine/arduino_and_MaxArm/MaxArm_ctl.cpp
/**
 * @file MaxArm_ctl.cpp
 * @author Wu TongXing
 * @brief Arduino与MaxArm通讯
 * @version 0.1
 * @date 2024-02-24
 */

#include "MaxArm_ctl.h"

static SoftwareSerial SerialM(rxPin, txPin);  //实例化软串口


/* CRC校验 */
static uint16_t checksum_crc8(const uint8_t *buf, uint16_t len)
{
    uint8_t check = 0;
    while (len--) {
        check = check + (*buf++);
    }
    check = ~check;
    return ((uint16_t) check) & 0x00FF;
}

MaxArm_ctl::MaxArm_ctl(void)
{
  
}


void MaxArm_ctl::serial_begin(void)
{
  SerialM.begin(9600);
}




//设置角度 0~240 (0~1000)
void MaxArm_ctl::set_angles(uint8_t* angles , uint16_t time)
{
  uint16_t pul[3];
  uint8_t msg[20] = {CONST_STARTBYTE1 , CONST_STARTBYTE2};
  msg[2] = FUNC_SET_ANGLE;
  msg[3] = 8;
  pul[0] = angles[0] > 240 ? 240 : angles[0];
  pul[1] = angles[1] > 240 ? 240 : angles[1];
  pul[2] = angles[2] > 240 ? 240 : angles[2];
  pul[0] = map(pul[0] , 0 , 240 , 0 , 1000);
  pul[1] = map(pul[1] , 0 , 240 , 0 , 1000);
  pul[2] = map(pul[2] , 0 , 240 , 0 , 1000);
  memcpy(&msg[4] , pul , 6);
  msg[10] = time & 0x00ff;
  msg[11] = (time>>8)&0x00FF;
  msg[12] = checksum_crc8((uint8_t*)&msg[2] , 10);
  SerialM.write((char*)&msg , 13);
}

//设置xyz
void MaxArm_ctl::set_xyz(int16_t* pos , uint16_t time)
{
  uint8_t msg[20] = {CONST_STARTBYTE1 , CONST_STARTBYTE2};
  msg[2] = FUNC_SET_XYZ;
  msg[3] = 8;
  memcpy(&msg[4] , pos , 6);
  msg[10] = time & 0x00ff;
  msg[11] = (time>>8)&0x00FF;
  msg[12] = checksum_crc8((uint8_t*)&msg[2] , 10);
  SerialM.write((char*)&msg , 13);
}

//设置末端舵机角度 0~180度
void MaxArm_ctl::set_pwmservo(uint8_t angle , uint16_t time)
{
  uint16_t pul = angle > 180 ? 180 : angle;
  uint8_t msg[10] = {CONST_STARTBYTE1 , CONST_STARTBYTE2};
  msg[2] = FUNC_SET_PWMSERVO;
  msg[3] = 4;
  pul = map(pul , 0 , 180 , 500, 2500);
  msg[4] = pul & 0x00ff;
  msg[5] = (pul>>8)&0x00FF;
  msg[6] = time & 0x00ff;
  msg[7] = (time>>8)&0x00FF;
  msg[8] = checksum_crc8((uint8_t*)&msg[2] , 6);
  SerialM.write((char*)&msg , 9);
}

//设置喷嘴功能
void MaxArm_ctl::set_SuctioNnozzle(int func)
{
  uint8_t msg[10] = {CONST_STARTBYTE1 , CONST_STARTBYTE2};
  msg[2] = FUNC_SET_SUCTIONNOZZLE;
  msg[3] = 1;
  // memcpy(&msg[4] , pul , 2);
  if(func < 0 && func > 3)
  {
    Serial.println("set_SuctioNnozzle ERROR");
    return;
  }
  msg[4] = func;
  msg[5] = checksum_crc8((uint8_t*)&msg[2] , 3);
  SerialM.write((char*)&msg , 6);
}

bool MaxArm_ctl::rec_handle(uint8_t* res , uint8_t func)
{
  int len = SerialM.available();
  // 限制读取的数据长度
  len = len > 30 ? 30 : len;
  uint8_t step = 0;
  uint8_t data[8] , index = 0;
  while(len--)
  {
    int rd = SerialM.read();
    switch(step)
    {
      case 0:
        if(rd == 0xAA)
        {
          step++;
        }
        break;
      case 1:
        if(rd == 0x55)
        {
          step++;
        }else{
          step = 0;
        }
        break;
      case 2:
        if(rd == func)
        {
          data[index++] = rd;
          step++;
        }else{
          step = 0;
        }
        break;
      case 3: //接收的数据长度必须为6个字节
        if(rd == 6)
        {
          data[index++] = rd;
          step++;
        }else{
          step = 0;
        }
        break;
      case 4:
        data[index++] = rd;
        if(index >= 8)
        {
          step++;
        }
        break;
      case 5:
        if(checksum_crc8(data , 8) == rd)
        {
          memcpy(res , &data[2] , 6);
          return true;
        }else{
          return false;
        }
        break;
      default:
        step = 0;
        break;
    }
  }
  return false;
}


//读取角度
bool MaxArm_ctl::read_angles(int* angles)
{
  uint8_t msg[10] = {CONST_STARTBYTE1 , CONST_STARTBYTE2};
  msg[2] = FUNC_READ_ANGLE;
  msg[3] = 0;
  msg[4] = checksum_crc8((uint8_t*)&msg[2] , 2);
  while(SerialM.available())
  {
    SerialM.read();
  }
  SerialM.write((char*)&msg , 5);
  uint16_t count = 3;
  delay(300);
  int16_t res[3];
  if(rec_handle((uint8_t*)res , 0x11))
  {
    angles[0] = map(res[0] , 0 , 1000 , 0 , 240);
    angles[1] = map(res[1] , 0 , 1000 , 0 , 240);
    angles[2] = map(res[2] , 0 , 1000 , 0 , 240);
    return true;
  }
  return false;
}

//读取xyz
bool MaxArm_ctl::read_xyz(int* pos)
{
  uint8_t msg[10] = {CONST_STARTBYTE1 , CONST_STARTBYTE2};
  msg[2] = FUNC_READ_XYZ;
  msg[3] = 0;
  msg[4] = checksum_crc8((uint8_t*)&msg[2] , 2);
  while(SerialM.available())
  {
    SerialM.read();
  }
  SerialM.write((char*)&msg , 5);
  delay(300);
  int16_t res[3];
  if(rec_handle((uint8_t*)res , 0x13))
  {
    pos[0] = res[0];
    pos[1] = res[1];
    pos[2] = res[2];
    return true;
  }
  return false;
}

File: 10. MaxArm Serial Communication/2. Underlying File Analysis/2.2 Master End/Raspberry Pi Communication/Raspberry Pi Routine/MaxArm_ctl.py
'''
@file:    MaxArm.py
@company: Hiwonder
@author:  CuZn
@date:    2024-02-26
@description: 接收 MaxArm消息 的通讯类
'''

#!/usr/bin/env python3
# encoding: utf-8
# MaxArm python sdk

import enum
import time
import struct
import serial

'''
0xAA 0x55 | func | data_len | data | check
'''

# CRC校验
def checksum_crc8(data , check_in = 0):
    check = check_in
    for b in data:
        check = check + b
    check = ~check
    return check & 0x00FF



# 功能号枚举
class PACKET_FUNCTION(enum.IntEnum): 
    FUNC_SET_ANGLE = 0x01
    FUNC_SET_XYZ = 0x03
    FUNC_SET_PWMSERVO = 0x05
    FUNC_SET_SUCTIONNOZZLE = 0x07
    FUNC_READ_ANGLE = 0x11
    FUNC_READ_XYZ = 0x13


'''
树莓派与MaxArm的通信类
'''
class MaxArm_ctl:
  
    # 构造函数
    def __init__(self, device = "/dev/ttyUSB0", baudrate=115200):
        self.__uart = serial.Serial(
                                        port=device,
                                        baudrate=baudrate,
                                        bytesize=serial.EIGHTBITS,
                                        parity=serial.PARITY_NONE,
                                        stopbits=serial.STOPBITS_ONE,
                                    )
    
    
    def map_func(self, value, from_min, from_max, to_min, to_max):
        # 将value从输入范围映射到输出范围
        return int((value - from_min) * (to_max - to_min) / (from_max - from_min) + to_min)
    
    
    '''
    用户使用函数 
    '''
    def serial_send(self , data):
        self.__uart.write(data)

    
    def set_angles(self, angles, time):
        # 将角度值映射到0~1000的范围
        mapped_angles = [self.map_func(angle, 0, 180, 0, 1000) for angle in angles]

        # 将角度值拆分为小端格式的两个字节
        angles_bytes = []
        for angle in mapped_angles:
            angle_bytes = struct.pack('<H', angle)
            angles_bytes.extend(angle_bytes)

        # 将时间拆分为小端格式的两个字节
        time_bytes = struct.pack('<H', time)

        # 构建要发送的数据帧
        data = bytearray([0xAA, 0x55, PACKET_FUNCTION.FUNC_SET_ANGLE, 0x08])
        data.extend(angles_bytes)  # 添加角度数据
        data.extend(time_bytes)    # 添加时间
        checksum = checksum_crc8(data[2:]) # 计算校验位
        data.append(checksum)  # 添加校验位

        # 发送数据帧
        self.serial_send(data)
        
        
    def set_xyz(self , pos , time):        
        # 将位置数据和时间打包为字节串
        pos_bytes = struct.pack('<hhh', *pos)
        time_bytes = struct.pack('<H', time)

        # 构建要发送的数据帧
        msg = bytearray([0xAA, 0x55, PACKET_FUNCTION.FUNC_SET_XYZ, 0x08])
        msg.extend(pos_bytes)
        msg.extend(time_bytes)
        
        # 计算校验位
        checksum = checksum_crc8(msg[2:]) # 计算校验位
        
        msg.append(checksum)

        # 发送数据帧
        self.serial_send(msg)
#        print(msg)


    def set_pwmservo(self , angle , time):
        angle = min(angle, 180)
        pul = self.map_func(angle, 0, 180, 500, 2500)

        # 构建要发送的数据帧
        data = bytearray([0xAA, 0x55, PACKET_FUNCTION.FUNC_SET_PWMSERVO, 0x04, pul & 0xFF, (pul >> 8) & 0xFF, time & 0xFF, (time >> 8) & 0xFF])

        # 计算校验位
        checksum = checksum_crc8(data[2:])
        data.append(checksum)

        # 发送数据帧
        self.serial_send(data)
#       print(data)


    def set_SuctioNnozzle(self , func):
        if func not in [1,2,3]:
            return
        # 构建要发送的数据帧
        data = bytearray([0xAA, 0x55, PACKET_FUNCTION.FUNC_SET_SUCTIONNOZZLE, 0x01, func & 0xFF])

        crc = checksum_crc8(data[2:])
        data.append(crc)
        self.serial_send(data)
#        print(data)
        
        
    def read_angles(self):
        # 构建要发送的读取角度的命令帧
        command = bytearray([0xAA, 0x55, 0x11, 0x00 , 0xEE])
        # 发送读取角度的命令帧
        self.serial_send(command)
        
        time.sleep(0.1)
        
        # 从串口接收数据
        response = self.__uart.read(12)
        
#        print(response)
        
        # 调用解析函数处理接收到的数据
        rec = self.rec_handle(response , 0x11)
        
        if rec:
            angles = struct.unpack('<hhh', rec)
        else:
            return
        return angles
        
    
    def read_xyz(self):
        # 构建要发送的读取角度的命令帧
        command = bytearray([0xAA, 0x55, 0x13, 0x00 , 0xEC])
        # 发送读取角度的命令帧
        self.serial_send(command)
        
        time.sleep(0.1)
        
        # 从串口接收数据
        response = self.__uart.read(11)
        
#        print(response)
        
        # 调用解析函数处理接收到的数据
        rec = self.rec_handle(response , 0x13)
        
        if rec:
            print(len(rec))
            xyz = struct.unpack('<hhh', rec)
        else:
            return
        return xyz


    def rec_handle(self , datas_in , func):
        step = 1
        len = 0
        rt_data = {0,0,0}
        count = 0
        
        btye_data = bytearray([])
        for data in datas_in:
            if step == 1:
                if data == 0xAA:
                    btye_data.clear()
                    step = 2
            elif step == 2:
                if data == 0x55:
                    step = 3
                else:
                    step = 1
            elif step == 3:
                if data == func:
                    btye_data.append(data)
                    step = 4
                else:
                    step = 1
            elif step == 4:
                if data == 6:
                    btye_data.append(data)
                    len = data
                    step = 5
                else:
                    step = 1
            elif step == 5:
                btye_data.append(data)
                len -= 1
                if len == 0:
                    step = 6
            elif step == 6:
                if checksum_crc8(btye_data) == data:
                    return btye_data[2:8]
                else:
                    return
                    
                    
                    
                



